src/Stratis.Bitcoin.Features.Consensus/Behaviors/ProvenHeadersReservedSlotsBehavior.cs

```diff
@@ -0,0 +1,150 @@
+using Stratis.Bitcoin.P2P.Protocol.Behaviors;
+using Stratis.Bitcoin.P2P.Protocol.Payloads;
+using Stratis.Bitcoin.Utilities;
+
+namespace Stratis.Bitcoin.Features.Consensus.Behaviors
+{
+   public class ProvenHeadersReservedSlotsBehavior : NetworkPeerBehavior
+   {
+      private readonly IConnectionManager connectionManager;
+      private readonly ILoggerFactory loggerFactory;
+      private readonly ConsensusSettings consensusSettings;
+      private readonly ICheckpoints checkpoints;
+      private readonly IChainState chainState;
+
+      /// <summary>
+      /// The proven header peers reserved slots threshold (%).
+      /// Represents the percentage of maximum connectable peers that we reserve for peers that are able to serve proven headers.
+      /// </summary>
+      private const decimal ProvenHeaderPeersReservedSlotsThreshold = 0.4M;
+
+      /// <summary>Instance logger.</summary>
+      private readonly ILogger logger;
+
+      public ProvenHeadersReservedSlotsBehavior(
+         IConnectionManager connectionManager,
+         ILoggerFactory loggerFactory,
+         ConsensusSettings consensusSettings,
+         ICheckpoints checkpoints,
+         IChainState chainState)
+      {
+         this.connectionManager = connectionManager;
+         this.loggerFactory = loggerFactory;
+         this.logger = loggerFactory.CreateLogger(this.GetType().FullName, $"[{this.GetHashCode():x}] ");
+         this.consensusSettings = consensusSettings;
+         this.checkpoints = checkpoints;
+         this.chainState = chainState;
+      }
+
+      /// <summary>
+      /// Processes and incoming message from the peer.
+      /// </summary>
+      /// <param name="peer">Peer from which the message was received.</param>
+      /// <param name="message">Received message to process.</param>
+      protected async Task OnMessageReceivedAsync(INetworkPeer peer, IncomingMessage message)
+      {
+         switch (message.Message.Payload)
+         {
+            case VersionPayload version:
+               await this.ProcessVersionAsync(peer, version).ConfigureAwait(false);
+               break;
+         }
+      }
+
+      /// <summary>
+      /// Processes "version" message received from the peer.
+      /// Ensures we leave some available connection slots to nodes that are enabled to serve Proven Headers, if the number of minimum PH enabled nodes connected isn't met.
+      /// </summary>
+      /// <param name="peer">Peer from which the message was received.</param>
+      /// <param name="version">Payload of "version" message to process.</param>
+      private Task ProcessVersionAsync(INetworkPeer peer, VersionPayload version)
+      {
+         bool isBelowLastCheckpoint = this.consensusSettings.UseCheckpoints && (this.chainState.ConsensusTip.Height <= this.checkpoints.GetLastCheckpointHeight());
+
+         // We enforce having free slots for PH enabled nodes only when we have passed last checkpoint.
+         if (!isBelowLastCheckpoint)
```

Comment:
I also noticed we don't check if PH is activated.

Reply:
👍 this logic should be executed when we are ahead of last checkpoint and ahead of PH activation height